sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQuerySourceBase.java

```diff
@@ -167,69 +182,70 @@ public void validate() {
     return BigQueryIO.getExtractFilePaths(extractDestinationDir, extractJob);
   }
 
-  List<BoundedSource<TableRow>> createSources(List<ResourceId> files, TableSchema tableSchema)
+  List<BoundedSource<T>> createSources(List<ResourceId> files, TableSchema tableSchema)
       throws IOException, InterruptedException {
-    final String jsonSchema = BigQueryIO.JSON_FACTORY.toString(tableSchema);
-
-    SerializableFunction<GenericRecord, TableRow> function =
-        new SerializableFunction<GenericRecord, TableRow>() {
-          private Supplier<TableSchema> schema = Suppliers.memoize(
-              Suppliers.compose(new TableSchemaFunction(), Suppliers.ofInstance(jsonSchema)));
-
-          @Override
-          public TableRow apply(GenericRecord input) {
-            return BigQueryAvroUtils.convertGenericRecordToTableRow(input, schema.get());
-          }};
 
-    List<BoundedSource<TableRow>> avroSources = Lists.newArrayList();
+    SerializableFunction<GenericRecord, T> function = parseFnFactory.apply(tableSchema);
```

Comment:
I'm wondering whether tableSchema is needed as a parameter at all. GenericRecord already has a Schema (getSchema()) - it's not the same as the TableSchema, but it probably contains all the same relevant information. I suspect in most cases users wouldn't use the schema at all because they know what fields they are querying for; and in cases where they don't, I think GenericRecord.getSchema() will be sufficient.

Reply:
Yeah, I could spend some time trying to get this to work.  I was trying to avoid having to re-implement the whole existing avro -> TableRow process here.

Reply:
So, please correct me if I'm wrong, but I don't think the avro scheme has enough information to maintain full backwards compatibility with the current implementation.

The big edge case I can think of right now is that timestamps in BigQuery are represented as `long`s in avro, and without the BigQuery schema, it's impossible to re-format them as dates and not just numbers (see BigQueryAvroUtils.java#204.

Reply:
Hm. I really don't want to subject users to having to implement a nested `SerializableFunction`. One slightly less ugly thing that comes to mind is suggest to create a helper `class BigQueryAvroRecord implements GenericRecord {TableSchema getBigQuerySchema();}` (which delegates the other methods of GenericRecord to the actual GenericRecord) and make the user pass a `SerializableFunction<BigQueryAvroRecord, T>`? Any obvious downsides to this? (I'm not  yet confident that this is actually good)

Reply:
I'm very torn on this in general.  On one hand having a higher-order function like this isn't very abnormal in other languages.  On the other hand Java makes this stupidly verbose, ugly, and hard to implement. 
Your proposal makes sense, but I'm not how sure how much I love squishing the BQ schema and avro record together like that.  If anything I'd prefer composition over inheritance, and have something like:
```
class SchemaAndRecord {
   private final GenericRecord record;
   private final TableSchema schema;
 
  // lots of java boilerplate.
}
```
But to me, this object model implies that the schema might change for each call and isn't constant.

Alternatively we could do something like `SerializableFunction2<TableSchema, GenericRecord, T>`, or like I suggested above something more specific to BigQuery like:
```
class ParseFn<T> extends Serializable { 
  T apply(TableSchema tableSchema, GenericRecord record) {...}
}
```
Again though this has similar problems as above.

Reply:
Composition is fine too. I agree that it's funky that it implies schema might change, but this is already true of the GenericRecord's `getSchema()` so I think adding another schema argument that doesn't really change doesn't make this worse.

To make this seem less horrible, consider that one day we'll probably add BigQueryIO.readAll() that reads a PCollection of table names or queries. Then the schema might actually be different, even though for any given schema there will be a lot of records with that schema.

Reply:
I'm sold :)

I'll cook something up.